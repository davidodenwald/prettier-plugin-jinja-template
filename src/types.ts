export interface Node {
  id: string;
  type: "root" | "expression" | "statement" | "block" | "comment" | "ignore";
  content: string;
  preNewLines: number;
  originalText: string;
  index: number;
  length: number;
  nodes: { [id: string]: Node };
}

type DelimiterChr = "" | "-" | "+";
export type Delimiter = {
  start: DelimiterChr;
  end: DelimiterChr;
};

export interface Expression extends Node {
  type: "expression";
  delimiter: Delimiter;
}

export interface Statement extends Node {
  type: "statement";
  keyword: string;
  delimiter: Delimiter;
}

export interface Block extends Node {
  type: "block";
  start: Statement;
  end: Statement;
  containsNewLines: boolean;
}

export type PlaceholderStyle = "jinja" | "json";
