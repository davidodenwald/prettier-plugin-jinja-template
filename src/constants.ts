export const Placeholder = {
  jinja: {
    startToken: "`~",
    endToken: "~`",
  },
  json: {
    startToken: "__~",
    endToken: "~__",
  },
} as const;

export const NOT_FOUND = -1 as const;
