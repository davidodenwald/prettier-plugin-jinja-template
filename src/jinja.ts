export const Placeholder = {
	startToken: "#~",
	endToken: "~#",
};

interface BaseNode {
	id: string;
	content: string;
	preNewLines: number;
	originalText: string;
	index: number;
	length: number;
	nodes: { [id: string]: Node };
}

export interface RootNode extends BaseNode {
	type: "root";
}

type DelimiterChr = "" | "-" | "+";
export type Delimiter = {
	start: DelimiterChr;
	end: DelimiterChr;
};

export interface ExpressionNode extends BaseNode {
	type: "expression";
	delimiter: Delimiter;
}

export interface StatementNode extends BaseNode {
	type: "statement";
	keyword: string;
	delimiter: Delimiter;
}

export interface BlockNode extends BaseNode {
	type: "block";
	start: StatementNode;
	end: StatementNode;
	containsNewLines: boolean;
}

export interface CommentNode extends BaseNode {
	type: "comment";
}

export interface IgnoreNode extends BaseNode {
	type: "ignore";
}

export type Node =
	| RootNode
	| ExpressionNode
	| StatementNode
	| BlockNode
	| CommentNode
	| IgnoreNode;
