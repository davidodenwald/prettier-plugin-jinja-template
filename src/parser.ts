import { Parser } from "prettier";
import {
	Delimiter,
	Node,
	Placeholder,
	Expression,
	Statement,
	Block,
	nonClosingStatements,
} from "./jinja";

const NOT_FOUND = -1;

const regex =
	/(?<node>{{(?<startDelimiterEx>[-+]?)\s*(?<expression>'([^']|\\')*'|"([^"]|\\")*"|[\S\s]*?)\s*(?<endDelimiterEx>[-+]?)}}|{%(?<startDelimiter>[-+]?)\s*(?<statement>(?<keyword>\w+)('([^']|\\')*'|"([^"]|\\")*"|[\S\s])*?)\s*(?<endDelimiter>[-+]?)%}|(?<comment>{#[\S\s]*?#})|(?<scriptBlock><(script)((?!<)[\s\S])*>((?!<\/script)[\s\S])*?{{[\s\S]*?<\/(script)>)|(?<styleBlock><(style)((?!<)[\s\S])*>((?!<\/style)[\s\S])*?{{[\s\S]*?<\/(style)>)|(?<ignoreBlock><!-- prettier-ignore-start -->[\s\S]*<!-- prettier-ignore-end -->))/;

export const parse: Parser<Node>["parse"] = (text) => {
	const statementStack: Statement[] = [];

	const root: Node = {
		id: "0",
		type: "root" as const,
		content: text,
		preNewLines: 0,
		originalText: text,
		index: 0,
		length: 0,
		nodes: {},
	};

	const generatePlaceholder = placeholderGenerator(text);

	let match;
	let i = 0;
	while ((match = text.slice(i).match(regex)) !== null) {
		if (!match.groups || match.index === undefined) {
			continue;
		}
		const matchLength = match[0].length;

		// skip script and style blocks
		if (match.groups.scriptBlock || match.groups.styleBlock) {
			i += match.index + matchLength;
			continue;
		}

		const matchText = match.groups.node;
		const expression = match.groups.expression;
		const statement = match.groups.statement;
		const ignoreBlock = match.groups.ignoreBlock;
		const comment = match.groups.comment;

		if (!matchText && !expression && !statement && !ignoreBlock && !comment) {
			continue;
		}
		const placeholder = generatePlaceholder();

		const emptyLinesBetween = text.slice(i, i + match.index).match(/^\s+$/) || [
			"",
		];
		const preNewLines = emptyLinesBetween.length
			? emptyLinesBetween[0].split("\n").length - 1
			: 0;

		const node = {
			id: placeholder,
			preNewLines,
			originalText: matchText,
			index: match.index + i,
			length: matchText.length,
			nodes: root.nodes,
		};

		if (comment || ignoreBlock) {
			root.content = root.content.replace(matchText, placeholder);
			root.nodes[node.id] = {
				...node,
				type: comment ? "comment" : "ignore",
				content: comment || ignoreBlock,
			};
		}

		if (expression) {
			const delimiter = (match.groups.startDelimiterEx ||
				match.groups.endDelimiterEx) as Delimiter;

			root.content = root.content.replace(matchText, placeholder);
			root.nodes[node.id] = {
				...node,
				type: "expression",
				content: expression,
				delimiter,
			} as Expression;
		}

		if (statement) {
			const keyword = match.groups.keyword;
			const delimiter = (match.groups.startDelimiter ||
				match.groups.endDelimiter) as Delimiter;

			if (nonClosingStatements.includes(keyword)) {
				root.content = root.content.replace(matchText, placeholder);
				root.nodes[node.id] = {
					...node,
					type: "statement",
					content: statement,
					keyword,
					delimiter,
				} as Statement;
			} else if (!keyword.startsWith("end")) {
				root.nodes[node.id] = {
					...node,
					type: "statement",
					content: statement,
					keyword,
					delimiter,
				} as Statement;
				statementStack.push(root.nodes[placeholder] as Statement);
			} else {
				let start: Statement | undefined;
				while (!start) {
					start = statementStack.pop();

					if (!start) {
						throw new Error(
							`No opening statement found for closing statement "${statement}".`
						);
					}

					const startKeyword = keyword.replace("end", "");
					if (startKeyword !== start.keyword) {
						if (start.keyword === "set") {
							root.content = root.content.replace(start.originalText, start.id);
							start = undefined;
							continue;
						}

						throw new Error(
							`Closing statement "${statement}" doesn't match Opening Statement "${start.content}".`
						);
					}
				}

				const end = {
					...node,
					type: "statement",
					content: statement,
					keyword,
					delimiter,
				} as Statement;
				root.nodes[end.id] = end;

				const blockText = root.content.slice(
					root.content.indexOf(start.originalText),
					root.content.indexOf(end.originalText) + end.length
				);

				const originalText = text.slice(start.index, end.index + end.length);
				const block = {
					id: generatePlaceholder(),
					type: "block",
					start: start,
					end: end,
					content: blockText.slice(start.length, blockText.length - end.length),
					preNewLines: start.preNewLines,
					containsNewLines: originalText.search("\n") !== NOT_FOUND,
					originalText,
					index: start.index,
					length: end.index + end.length - start.index,
					nodes: root.nodes,
				} as Block;
				root.nodes[block.id] = block;

				root.content = root.content.replace(blockText, block.id);
			}
		}

		i += match.index + matchLength;
	}

	for (const stmt of statementStack) {
		if (stmt.keyword === "set") {
			root.content = root.content.replace(stmt.originalText, stmt.id);
		} else {
			throw new Error(
				`No closing statement found for opening statement "${stmt.content}".`
			);
		}
	}

	return root;
};

const placeholderGenerator = (text: string) => {
	let id = 0;

	return (): string => {
		while (true) {
			id++;

			const placeholder = Placeholder.startToken + id + Placeholder.endToken;
			if (!text.includes(placeholder)) {
				return placeholder;
			}
		}
	};
};
