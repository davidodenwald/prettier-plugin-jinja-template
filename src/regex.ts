/*
Do not edit this file!!!

The file "src/regex_editable.ts" contains the source for this regex and is transformed by "babel-plugin-transform-regex".
Run "npm run build:regex" to generate this file again afer editing the source.
*/
export default /(?<node>\{\{(?<startDelimiterEx>(?:[\-+])?)\s*(?<expression>(?:(?=((?:'[^']*'|"[^"]*")|(?:[\s\S])))\4)*?)\s*(?<endDelimiterEx>(?:[\-+])?)\}\}|\{%(?<startDelimiter>(?:[\-+])?)\s*(?<statement>(?<keyword>\w+)(?:(?=((?:'[^']*'|"[^"]*")|(?:[\s\S])))\9)*?)\s*(?<endDelimiter>(?:[\-+])?)%\}|(?<ignoreBlock>(?:<!--\s*prettier-ignore-start\s*-->|\{#\s*prettier-ignore-start\s*#\})(?:(?:[\s\S])*?)(?:<!--\s*prettier-ignore-end\s*-->|\{#\s*prettier-ignore-end\s*#\}))|(?<comment>\{#(?:(?:[\s\S])*?)#\})|(?<scriptBlock><script(?:(?:[\s\S])*?)>(?:(?:[\s\S])*?)<\/script>\s*)|(?<styleBlock><style(?:(?:[\s\S])*?)>(?:(?:[\s\S])*?)<\/style>\s*))/u;
