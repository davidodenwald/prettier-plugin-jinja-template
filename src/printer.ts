import { AstPath, Doc, Options, Printer } from "prettier";
import { builders, utils } from "prettier/doc";
import { extendedOptions } from "./index";
import { Block, Expression, Node, Placeholder, Statement } from "./jinja";

const NOT_FOUND = -1;

process.env.PRETTIER_DEBUG = "true";

export const getVisitorKeys = (
	ast: Node | { [id: string]: Node },
): string[] => {
	if ("type" in ast) {
		return ast.type === "root" ? ["nodes"] : [];
	}
	return Object.values(ast)
		.filter((node) => {
			return node.type === "block";
		})
		.map((e) => e.id);
};

export const print: Printer<Node>["print"] = (path) => {
	const node = path.getNode();
	if (!node) {
		return [];
	}

	switch (node.type) {
		case "expression":
			return printExpression(node as Expression);
		case "statement":
			return printStatement(node as Statement);
		case "comment":
			return printCommentBlock(node);
		case "ignore":
			return printIgnoreBlock(node);
	}
	return [];
};

const printExpression = (node: Expression): builders.Doc => {
	const multiline = node.content.includes("\n");

	const expression = builders.group(
		builders.join(" ", [
			["{{", node.delimiter],
			multiline
				? builders.indent(getMultilineGroup(node.content))
				: node.content,
			multiline
				? [builders.hardline, node.delimiter, "}}"]
				: [node.delimiter, "}}"],
		]),
		{
			shouldBreak: node.preNewLines > 0,
		},
	);

	return node.preNewLines > 1
		? builders.group([builders.trim, builders.hardline, expression])
		: expression;
};

const printStatement = (node: Statement): builders.Doc => {
	const multiline = node.content.includes("\n");

	const statemnt = builders.group(
		builders.join(" ", [
			["{%", node.delimiter],
			multiline
				? builders.indent(getMultilineGroup(node.content))
				: node.content,
			multiline
				? [builders.hardline, node.delimiter, "%}"]
				: [node.delimiter, "%}"],
		]),
		{ shouldBreak: node.preNewLines > 0 },
	);

	if (
		["else", "elif"].includes(node.keyword) &&
		surroundingBlock(node)?.containsNewLines
	) {
		return [builders.dedent(builders.hardline), statemnt, builders.hardline];
	}
	return statemnt;
};

const printCommentBlock = (node: Node): builders.Doc => {
	const comment = builders.group(node.content, {
		shouldBreak: node.preNewLines > 0,
	});

	return node.preNewLines > 1
		? builders.group([builders.trim, builders.hardline, comment])
		: comment;
};

const printIgnoreBlock = (node: Node): builders.Doc => {
	return node.content;
};

export const embed: Printer<Node>["embed"] = () => {
	return async (
		textToDoc: (text: string, options: Options) => Promise<Doc>,
		print: (
			selector?: string | number | Array<string | number> | AstPath,
		) => Doc,
		path: AstPath,
		options: Options,
	): Promise<Doc | undefined> => {
		const node = path.getNode();
		if (!node || !["root", "block"].includes(node.type)) {
			return undefined;
		}

		const mapped = await Promise.all(
			splitAtElse(node).map(async (content) => {
				let doc;
				if (content in node.nodes) {
					doc = content;
				} else {
					/**
					 * The lwc parser is the same as the "html" parser,
					 * but also formats LWC-specific syntax for unquoted template attributes.
					 */
					const parser = (options as extendedOptions).quoteAttributes
						? "html"
						: "lwc";

					doc = await textToDoc(content, {
						...options,
						parser,
					});
				}

				let ignoreDoc = false;

				return utils.mapDoc(doc, (currentDoc) => {
					if (typeof currentDoc !== "string") {
						return currentDoc;
					}

					if (currentDoc === "<!-- prettier-ignore -->") {
						ignoreDoc = true;
						return currentDoc;
					}

					const idxs = findPlaceholders(currentDoc).filter(
						([start, end]) => currentDoc.slice(start, end + 1) in node.nodes,
					);
					if (!idxs.length) {
						ignoreDoc = false;
						return currentDoc;
					}

					const res: builders.Doc = [];
					let lastEnd = 0;
					for (const [start, end] of idxs) {
						if (lastEnd < start) {
							res.push(currentDoc.slice(lastEnd, start));
						}

						const p = currentDoc.slice(start, end + 1) as string;

						if (ignoreDoc) {
							res.push(node.nodes[p].originalText);
						} else {
							res.push(path.call(print, "nodes", p));
						}

						lastEnd = end + 1;
					}

					if (lastEnd > 0 && currentDoc.length > lastEnd) {
						res.push(currentDoc.slice(lastEnd));
					}

					ignoreDoc = false;
					return res;
				});
			}),
		);

		if (node.type === "block") {
			const block = buildBlock(path, print, node as Block, mapped);

			return node.preNewLines > 1
				? builders.group([builders.trim, builders.hardline, block])
				: block;
		}
		return [...mapped, builders.hardline];
	};
};

const getMultilineGroup = (content: String): builders.Group => {
	// Dedent the content by the minimum indentation of any non-blank lines.
	const lines = content.split("\n");
	const minIndent = Math.min(
		...lines
			.slice(1) // can't be the first line
			.filter((line) => line.trim())
			.map((line) => line.search(/\S/)),
	);

	return builders.group(
		lines.map((line, i) => [
			builders.hardline,
			i === 0
				? line.trim() // don't dedent the first line
				: line.trim()
					? line.slice(minIndent).trimEnd()
					: "",
		]),
	);
};

const splitAtElse = (node: Node): string[] => {
	const elseNodes = Object.values(node.nodes).filter(
		(n) =>
			n.type === "statement" &&
			["else", "elif"].includes((n as Statement).keyword) &&
			node.content.search(n.id) !== NOT_FOUND,
	);
	if (!elseNodes.length) {
		return [node.content];
	}

	const re = new RegExp(`(${elseNodes.map((e) => e.id).join(")|(")})`);
	return node.content.split(re).filter(Boolean);
};

/**
 * Returns the indexs of the first and the last character of any placeholder
 * occuring in a string.
 */
export const findPlaceholders = (text: string): [number, number][] => {
	const res = [];
	let i = 0;

	while (true) {
		const start = text.slice(i).search(Placeholder.startToken);
		if (start === NOT_FOUND) break;
		const end = text
			.slice(start + i + Placeholder.startToken.length)
			.search(Placeholder.endToken);
		if (end === NOT_FOUND) break;

		res.push([
			start + i,
			end + start + i + Placeholder.startToken.length + 1,
		] as [number, number]);
		i += start + Placeholder.startToken.length;
	}
	return res;
};

export const surroundingBlock = (node: Node): Block | undefined => {
	return Object.values(node.nodes).find(
		(n) => n.type === "block" && n.content.search(node.id) !== NOT_FOUND,
	) as Block;
};

const buildBlock = (
	path: AstPath<Node>,
	print: (path: AstPath<Node>) => builders.Doc,
	block: Block,
	mapped: (string | builders.Doc[] | builders.DocCommand)[],
): builders.Doc => {
	// if the content is empty or whitespace only.
	if (block.content.match(/^\s*$/)) {
		return builders.fill([
			path.call(print, "nodes", block.start.id),
			builders.softline,
			path.call(print, "nodes", block.end.id),
		]);
	}
	if (block.containsNewLines) {
		return builders.group([
			path.call(print, "nodes", block.start.id),
			builders.indent([builders.softline, mapped]),
			builders.hardline,
			path.call(print, "nodes", block.end.id),
		]);
	}
	return builders.group([
		path.call(print, "nodes", block.start.id),
		mapped,
		path.call(print, "nodes", block.end.id),
	]);
};
