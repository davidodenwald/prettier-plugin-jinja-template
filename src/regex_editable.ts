import { regex } from "regex";

export default regex`
	(?<node>
		# Expression
		\{\{
			(?<startDelimiterEx>\g<DELIMITER>?)\s*
				(?<expression>(?>\g<ESCAPEQUOTES> | \g<ALLSYMBOLS>)*?)
			\s*(?<endDelimiterEx>\g<DELIMITER>?)
		\}\}
		|
		# Statement
		\{%
			(?<startDelimiter>\g<DELIMITER>?)\s*
				(?<statement>
					(?<keyword>\w+)
					(?>\g<ESCAPEQUOTES> | \g<ALLSYMBOLS>)*?
				)
			\s*(?<endDelimiter>\g<DELIMITER>?)
		%\}
		|
		# Ignore block
		(?<ignoreBlock>
			(<!--\s*prettier-ignore-start\s*--> | \{\#\s*prettier-ignore-start\s*\#\})
				\g<EVERYTHING>
			(<!--\s*prettier-ignore-end\s*--> | \{\#\s*prettier-ignore-end\s*\#\})
		)
		|
		# Comment
		(?<comment>\{\#\g<EVERYTHING>\#\})
		|
		# Script Block
		(?<scriptBlock><script\g<EVERYTHING>>\g<EVERYTHING><\/script>\s*)
		|
		# Style Block
		(?<styleBlock><style\g<EVERYTHING>>\g<EVERYTHING><\/style>\s*)
	)

	(?(DEFINE)
		(?<EVERYTHING>		\g<ALLSYMBOLS>*?)
		(?<ALLSYMBOLS>		[\s\S])
		(?<DELIMITER>			[\-+])
		(?<ESCAPEQUOTES>	'[^']*'|"[^"]*")
	)
`;
