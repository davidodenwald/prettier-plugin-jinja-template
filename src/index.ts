import { Node } from "./jinja";
import { parse } from "./parser";
import { print, embed, getVisitorKeys } from "./printer";
import { Parser, Printer, SupportLanguage } from "prettier";

const PLUGIN_KEY = "jinja-template";

export const languages: SupportLanguage[] = [
	{
		name: "JinjaTemplate",
		parsers: [PLUGIN_KEY],
		extensions: [".jinja", ".jinja2", ".j2", ".html"],
		vscodeLanguageIds: ["jinja"],
	},
];

export const parsers = {
	[PLUGIN_KEY]: <Parser<Node>>{
		astFormat: PLUGIN_KEY,
		parse,
		locStart: (node) => node.index,
		locEnd: (node) => node.index + node.length,
	},
};

export const printers = {
	[PLUGIN_KEY]: <Printer<Node>>{
		print,
		embed,
		getVisitorKeys,
	},
};
