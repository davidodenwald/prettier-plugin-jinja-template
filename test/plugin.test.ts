import { readdirSync, readFileSync } from "fs";
import { join } from "path";
import { format } from "prettier/standalone";

import { plugin, PLUGIN_KEY } from "../src";

describe("plugins", () => {

	const formated = (input: string) => format(input, {
		parser: PLUGIN_KEY,
		plugins: [plugin]
	});

	describe("valid", () => {
		const validCaseFolder = join(__dirname, "cases");
		const tests = readdirSync(validCaseFolder).filter((f) => f.endsWith(".jinja"));

		it.each(tests)("should json jinja: %s be valid", async (test) => {
			const testPath = join(validCaseFolder, test);
			const input = readFileSync(testPath).toString();

			expect(await formated(input)).toMatchSnapshot(test);
		});
	});

	describe("broken", () => {

		const brokenNotOpenStatementFolder = join(__dirname, "cases", "broken_not_open_statement");
		const testsBroken = readdirSync(brokenNotOpenStatementFolder).filter((f) => f.endsWith(".jinja"));

		it.each(testsBroken)("should json jinja: %s be broken_not_open_statement", async (test) => {
			const testPath = join(brokenNotOpenStatementFolder, test);
			const input = readFileSync(testPath).toString();

			await expect(() => formated(input)).rejects.toThrow(/No opening statement found for closing statement "(.*)"/);
		});

	});
});
