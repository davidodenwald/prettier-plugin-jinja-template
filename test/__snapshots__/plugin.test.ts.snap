// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugins valid should json jinja: collition.jinja be valid: collition.jinja 1`] = `
"#~1~# {{ i }}
"
`;

exports[`plugins valid should json jinja: comment.jinja be valid: comment.jinja 1`] = `
"{#
	Language will be used as ab_variant shortcut
	tracking callback is our own patched callback
  
	Integration note:
  
	window.emos3 = {
	  stored: [],
	  send: function(p){this.stored.push(p);}
	};

	This script tag must be moved to the head tag so
	that onclick events can already use the send method.
#}
"
`;

exports[`plugins valid should json jinja: empty.jinja be valid: empty.jinja 1`] = `""`;

exports[`plugins valid should json jinja: expression.jinja be valid: expression.jinja 1`] = `
"[{{ user.name }}, "{{ user.age }}"]
"
`;

exports[`plugins valid should json jinja: expression_2.jinja be valid: expression_2.jinja 1`] = `
"{"userName: "~{{ user.id }}~{{ user.name }}~"}
"
`;

exports[`plugins valid should json jinja: expression_escaped.jinja be valid: expression_escaped.jinja 1`] = `
"{{ '{{' }}
{{ '}}' }}
"
`;

exports[`plugins valid should json jinja: expression_long.jinja be valid: expression_long.jinja 1`] = `
"{
  "long": "{{ really really looooooooooooooooooooooooooooooooooooooooong fdsf sdf sdf sdf sdf sdf sdf dsfexpression that should get a own line }}"
}
"
`;

exports[`plugins valid should json jinja: expression_multiline.jinja be valid: expression_multiline.jinja 1`] = `
"{{
  {
    'dict': 'of',
    'key': 'and',
    'value': 'pairs'
  }
}}
"
`;

exports[`plugins valid should json jinja: expression_whitespace.jinja be valid: expression_whitespace.jinja 1`] = `
"{
  "info": {{- some_variable -}},
  "obj": {{ some_variable -}},
  "var": {{- some_variable }}
}
"
`;

exports[`plugins valid should json jinja: issue_25.jinja be valid: issue_25.jinja 1`] = `
"{% if true %}
  {% if true %}
  {% endif %}
{% endif %}
"
`;

exports[`plugins valid should json jinja: statement.jinja be valid: statement.jinja 1`] = `
"[
  {% for item in seq %}
    {{ item }},

  {% endfor %}
]
"
`;

exports[`plugins valid should json jinja: statement_broken_2.jinja be valid: statement_broken_2.jinja 1`] = `
"[
{% if title %}{"title": {{ title }}}]
"
`;

exports[`plugins valid should json jinja: statement_empty_block.jinja be valid: statement_empty_block.jinja 1`] = `
"{% block title %}
{% endblock %}
"
`;

exports[`plugins valid should json jinja: statement_long.jinja be valid: statement_long.jinja 1`] = `
"{
  "captions": [
    {%
      for href, caption in [
        ('index.html', 'Index'),
        ('about.html', 'About'),
        ('downloads.html', 'Downloads')
      ]
    %}
      {"href": {{ href }}, "caption": {{ caption }}}
    {% endfor %}
  ]
}
"
`;

exports[`plugins valid should json jinja: statement_multiple.jinja be valid: statement_multiple.jinja 1`] = `
"{
{% for row in database %}
  "ol":{
{% for element in row %}
    {% if element.active %}
      "li":{
        "description": {{ element.description }}
      }
    {% endif %}
  {% endfor %}}
{% endfor %}
"arr":[{{ row[0] }},{{ row[1] }},{{ row[2] }},{{ row[3] }}, {{ row[3] }}]
}
"
`;

exports[`plugins valid should json jinja: statement_multiple_else.jinja be valid: statement_multiple_else.jinja 1`] = `
"{% if True %}
  {"o": {"a": {{ abc }}}}
{% endif %}
{% if True %}
  {"b": {{ abc }}}
{% else %}
  {"p":{"e": {{ abc }},{% if False %}"z":{{ bcd }}{% else %}"x": {{ cde }}{% endif %}}
{% endif %}

"
`;

exports[`plugins valid should json jinja: statement_unknown.jinja be valid: statement_unknown.jinja 1`] = `
"{% foo %}
  {"h1": "Hello"}
{% endfoo %}
"
`;

exports[`plugins valid should json jinja: statement_whitespace.jinja be valid: statement_whitespace.jinja 1`] = `
"[
  {%- for item in seq -%}
    {"item": {{ item }}}
  {%- endfor %}
]
"
`;

exports[`plugins valid should json jinja: statement_with.jinja be valid: statement_with.jinja 1`] = `
"{{ obj.method("argument")|default("", True) }}
{% with text='text', version=1 %}
  {% include 'directory/template.jinja2' %}
{% endwith %}
"
`;
